nexus/src/db/queries/network_interface.rs

```diff
@@ -979,16 +989,342 @@ impl QueryFragment<Pg> for InsertNetworkInterfaceQueryValues {
+        out.push_bind_param::<sql_types::Uuid, Uuid>(&self.interface_id)?;
+        out.push_sql(" AND ");
+        out.push_identifier(dsl::time_deleted::NAME)?;
+        out.push_sql(" IS NULL) OR (SELECT COUNT(1) FROM ");
+        NETWORK_INTERFACE_FROM_CLAUSE.walk_ast(out.reborrow())?;
+        out.push_sql(" WHERE ");
+        out.push_identifier(dsl::instance_id::NAME)?;
+        out.push_sql(" = ");
+        out.push_bind_param::<sql_types::Uuid, Uuid>(&self.instance_id)?;
+        out.push_sql(" AND ");
+        out.push_identifier(dsl::time_deleted::NAME)?;
+        out.push_sql(" IS NULL) = 1, ");
+        out.push_bind_param::<sql_types::Text, String>(&self.instance_id_str)?;
+        out.push_sql(", ");
+        out.push_bind_param::<sql_types::Text, &str>(
+            &DeleteNetworkInterfaceError::HAS_SECONDARIES_SENTINEL,
+        )?;
+        out.push_sql(") AS UUID)), instance AS MATERIALIZED (SELECT CAST(IF(EXISTS(SELECT ");
+        out.push_identifier(db::schema::instance::dsl::id::NAME)?;
+        out.push_sql(" FROM ");
+        INSTANCE_FROM_CLAUSE.walk_ast(out.reborrow())?;
+        out.push_sql(" WHERE ");
+        out.push_identifier(db::schema::instance::dsl::id::NAME)?;
+        out.push_sql(" = ");
+        out.push_bind_param::<sql_types::Uuid, Uuid>(&self.instance_id)?;
+        out.push_sql(" AND ");
+        out.push_identifier(db::schema::instance::dsl::time_deleted::NAME)?;
+        out.push_sql(" IS NULL AND ");
+        out.push_identifier(db::schema::instance::dsl::state::NAME)?;
+        out.push_sql(" = ");
+        out.push_bind_param::<db::model::InstanceStateEnum, db::model::InstanceState>(&self.instance_state)?;
+        out.push_sql("), ");
+        out.push_bind_param::<sql_types::Text, String>(&self.instance_id_str)?;
+        out.push_sql(", ");
+        out.push_bind_param::<sql_types::Text, &str>(
+            &DeleteNetworkInterfaceError::INSTANCE_RUNNING_SENTINEL,
+        )?;
+        out.push_sql(") AS UUID))");
+        out.push_sql(" UPDATE ");
+        NETWORK_INTERFACE_FROM_CLAUSE.walk_ast(out.reborrow())?;
+        out.push_sql(" SET ");
+        out.push_identifier(dsl::time_deleted::NAME)?;
+        out.push_sql(" = NOW() WHERE ");
+        out.push_identifier(dsl::id::NAME)?;
+        out.push_sql(" = ");
+        out.push_bind_param::<sql_types::Uuid, Uuid>(&self.interface_id)?;
+        out.push_sql(" AND ");
+        out.push_identifier(dsl::time_deleted::NAME)?;
+        out.push_sql(" IS NULL");
+        Ok(())
+    }
+}
+
+impl RunQueryDsl<DbConnection> for DeleteNetworkInterfaceQuery {}
+
+/// Errors related to deleting a network interface from an instance
+#[derive(Debug)]
+pub enum DeleteNetworkInterfaceError {
+    /// Attempting to delete the primary interface, while there still exist
+    /// secondary interfaces.
+    InstanceHasSecondaryInterfaces(Uuid),
+    /// Instance must be stopped prior to deleting interfaces from it
+    InstanceMustBeStopped(Uuid),
```

Comment:
I don't see any tests which trigger these results in this module - would it be hard to add them?

Reply:
There are checks for these cases in [the integration tests](https://github.com/oxidecomputer/omicron/pull/1143/files#diff-8cbc06849f2b6c581df600a3f06e0687c983a1be354633bfa675441410ed0387) now. That includes trying to create / delete NICs from stopped instances, deleting the primary before the secondary, etc.